include/xgboost/c_api.h

```diff
@@ -128,10 +128,11 @@ XGB_DLL int XGBGetGlobalConfig(char const **out_config);
  * \brief load a data matrix
  * \param fname the name of the file
  * \param silent whether print messages during loading
+ * \param dsplit data split mode
  * \param out a loaded data matrix
  * \return 0 when success, -1 when failure happens
  */
-XGB_DLL int XGDMatrixCreateFromFile(const char *fname, int silent, DMatrixHandle *out);
+XGB_DLL int XGDMatrixCreateFromFile(const char *fname, int silent, int dsplit, DMatrixHandle *out);
```

Comment:
This is the only place I'm aware of that's dealing with data split. Do those third party libraries support splitting?

If we take a concrete example, say, the Higgs dataset, here are the steps to get to training:
* Download the dataset from UCI.
* Unzip to csv.
* Manually split the csv to train/test (e.g. 90%/10%).

Current behavior:
* Local training: feed the train/test csv files to `DMatrix`, not additional splitting happens.
* Distributed training: feed the train/test csv files to `DMatrix`, xgboost automatically splits both files by the number of workers.
* Federated learning: we need to manually split the train/test files first to simulate the federated environment. For each worker, feed the already split train/test files to `DMatrix`, no additional splitting happens.

With this PR:
* Local training: no behavior change.
* Distributed training: feed the train/test csv files to `DMatrix`, xgboost automatically splits both files by the number of workers. There is an additional option to manually split the files first, then create the `DMatrix` with `data_split_mode='none'`.
* Federated learning: no behavior change.

Future:
* Local training: no change.
* Distributed training: support column-wise split by passing in `data_split_mode='col'` to the `DMatrix`.
* Federated learning: support vertical learning by passing in, say, `federated_data_partition=vertical` to the communicator. We'd assume the `DMatrix` is still created with `data_split_mode='none'`.

Does that make more sense?

I'll add a new API to take in JSON and support the data split mode.

Reply:
Thank you for the explanation. My concern of using the file parser is that I think we will eventually integrate xgboost with some sort of federated learning framework that manages the data just like dask/spark for distributed computing. In that case, the file parser route is less ideal.

Reply:
Thought about this more. I think we do have two separate concepts mixed together. Perhaps it's better to separate them out. Let's look at each training scenario.

**Local training**: single input only.

**Distributed training with non-file api**: assuming we are using a non-file api (spark/cudf etc.), and data is already split during preprocessing. We still need to tell xgboost how the data is split. We can add a field to `MetaInfo`, call it `data_distribution`? Which defaults to `row`, but can also be `column`, maybe later `block`.

**Federated learning**: here data is already siloed between workers, so we can rely on the same `MetaInfo.data_distribution` field.

Finally, the **`XGDMatrixCreateFromFile` API**: probably not very useful for production, but helpful for quick benchmarking and ease of use (just point `DMatrix` at a file). For local and federated training, it's single input only. For distributed training, currently by default it splits the input by row. We want to be able to pass in a parameter so it splits by column. We can reuse the existing `data_split_mode` enum.

In summary, here is the proposal:
* Add a `data_distribution` field to `MetaInfo` with values:
  - row (default)
  - column
  - block (for future use; can leave it out for now)
* Deprecate the current c api  `XGDMatrixCreateFromFile`.
* Add a new c api `XGDMatrixCreateFromFile` which takes in a JSON config that includes the `data_split_mode` field.
* Don't change jvm/R APIs for now, but add a new `data_distribution` parameter to the python `DMatrix` class.

Does that sound reasonable?

@hcho3 since he worked on block distribution.

Reply:
I think adding a parameter to DMatrix is possible but would be great if it's limited to the file API and only at C API level (instead of an attribute of DMatrix). For instance, it can be a parameter in the input URI:
``` python
DMatrix("data.csv?need_split=1", data_split_mode="column")
```

Reply:
But we still need an indicator somewhere to know how the data is split, since the training behavior is different for different split modes.

Reply:
We can let `data_split_mode` be the indicator for how the data is split, and use the `need_split` in the URI to tell DMatrix whether it should slice itself.

Reply:
But ultimately `data_split_mode` needs to be in `MetaInfo`, right? With your example
```
DMatrix("data.csv?need_split=1", data_split_mode="column")
```
don't we still need to pass in the `data_split_mode` through the c api?

Reply:
Yup, apologies. We still need the deprecation

Reply:
What do you think? ;-)

Reply:
Sounds good. We can probably just pass `need_split` as part of the JSON request.

Reply:
Added `XGDMatrixCreateFromFileV2` for lack of a better name. Ended up not exposing `need_split` and just replicating the current behavior.

Reply:
Ok added `need_split` to the JSON config, but for now it's not exposed in the python api. We can always add it later if needed.